### PR TITLE
Double gyre example

### LINKED ISSUE #678: Double gyre example
Would be a good example to showcase the "Box" model with 3 `Bounded` directions.

cc @navidcy

### PR DESCRIPTION
**Pull request by navidcy**
This needs work but following suggestion, I submit a PR and we all work together... :)

Closes #678.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
This would make an awesome example!

By modifying the parameters it could also be used at https://github.com/CliMA/OceanModelComparison.jl to compare against the double gyre solution produced by ClimateMachine.jl.

**Comment by glwagner** [comment]
> By modifying the parameters it could also be used at https://github.com/CliMA/OceanModelComparison.jl to compare against the double gyre solution produced by ClimateMachine.jl.

Indeed. I also want to see if I can get some eddies and do some time-averaging of the barotropic flow for pedagogical purposes.

**Comment by navidcy** [comment]
Any chance we use vertically stretched grid?

**Comment by glwagner** [comment]
Stretched grid implementation isn't complete...

**Comment by navidcy** [comment]
Hint, hint?

**Comment by glwagner** [comment]
We can't use it; it doesn't work.

**Comment by francispoulin** [comment]
Does this work without a vertically stretched grid?

**Comment by glwagner** [comment]
Yes, a double gyre simulation can be run on a regular grid.

**Comment by navidcy** [comment]
@francispoulin, yeah, why not? :) One might require more vertical grid points or what not, but I don't see any reason why it shouldn't.

**Comment by francispoulin** [comment]
Thanks.  It seemed that was on the wish list but wasn't sure if anything bad was happening without the stretched grid.

I will hopefully run this example sometime soon and see the resutls myself.

**Comment by navidcy** [comment]
It's been there but we haven't quite yet found the parameters that would make this work and still be relatively "cheap" to add as an example in the Docs that would run automatically with Literate.jl
:)

**Comment by glwagner** [comment]
Last we talked, we decided that the most appropriate way forward with this example would be to pick it up from a checkpoint, and add the checkpointed data to the repository. This would make it so the example could be run in a reasonable amount of time, and also provide an example of checkpointing.

**Comment by navidcy** [comment]
We should resurrect now with GPU on triply-Bounded topology + VerticallyStretchedGrid. :)

**Comment by ali-ramadhan** [comment]
Out of curiousity, is the plan for this PR to add an example as an eddying baroclinic double gyre on the GPU with `IncompressibleModel` while #1433 adds a viscous barotropic double gyre with `HydrostaticFreeSurfaceModel` as a validation experiment?

**Comment by glwagner** [comment]
There's no plan! We should make one.

**Comment by francispoulin** [comment]
> Out of curiousity, is the plan for this PR to add an example as an eddying baroclinic double gyre on the GPU with `IncompressibleModel` while #1433 adds a viscous barotropic double gyre with `HydrostaticFreeSurfaceModel` as a validation experiment?

As a first step why not set up both `IncompressibleModel` and `HydrostaticFreeSurfaceModel` for a barotropic double gyre?  Since there will be much in the way of nonhydrostatic effects, it would be a way to compare the two models in terms of the accuracy and the efficiency.  Then it should be easy to change both to do a baroclinic double gyre.

**Comment by navidcy** [comment]
@glwagner, this should be feasible now with the but in the `VerticallyStretchedRectilinearGrid` and also with the `HydrostaticFreeSurfaceModel` with `ImplicitFreeSurface()`.

I made some changes. Should we zoom and finish it off?

**Comment by glwagner** [comment]
> @glwagner, this should be feasible now with the but in the `VerticallyStretchedRectilinearGrid` and also with the `HydrostaticFreeSurfaceModel` with `ImplicitFreeSurface()`.
> 
> I made some [changes](https://github.com/CliMA/Oceananigans.jl/pull/1085/commits/5fa8ff5ff44d357be989e68350537f11b643bb12). Should we zoom and finish it off?

This would be nice to have. Let's do it.

Should we run a short spin-up, or illustrate checkpointing with an equilibrated solution?

**Comment by navidcy** [comment]
Yes. But before doing that let's make sure it runs! It was blowing up when I was trying before. I may have done something wrong with forcing/bcs?

**Comment by glwagner** [comment]
> Yes. But before doing that let's make sure it runs! It was blowing up when I was trying before. I may have done something wrong with forcing/bcs?

That's a good starting point...

**Comment by navidcy** [comment]
Some progress here...

https://user-images.githubusercontent.com/7112768/151718447-2a5fd619-3c4c-43c8-9aa5-29c97e0129b4.mp4

**Comment by navidcy** [comment]
double_gyre_circulation

**Comment by navidcy** [comment]
We need to think how we do this. We need a long run to produce a good figure. E.g., 100 years?

Should we run this somewhere and save output and then in this example show how we can load the output and plot?

Or should we run this somewhere for 95 years and save a checkpoint and then use this example to demonstrate how we can restart from a checkpoint and run for 5 years to produce animations/plots?

**Comment by navidcy** [comment]
If we use checkpointer then the PR will close #1136.

**Comment by francispoulin** [comment]
Great to see you working on this @navidcy !  

Some questions/comments:

I think having an animation is a great idea.  Any particular reason you are plotting `u`?  Traditionally, people plot the streamfunction (pressure) and or the voriticty.  What do you think of including the streamfunction in the animation?

From looking at the animation I was surprised that we don't see a double gyre.  But from your next image we see there is a double gyre in the barotropic streamfunction. This makes me think that if we plot the total streamfunction (or maybe even the barotropic and baroclinic components) we should get images that are more like what I've seen in papers.

How long did this take you to run and what resolution did you use?

**Comment by glwagner** [comment]
I think we should run it at a reasonably high resolution to produce eddies, and then checkpoint the run and upload it to `OceananigansArtifacts.jl`. Then we can add the `datadep` to Oceananigans source, download the checkpoint in the example, restart from there and run maybe 100 iterations to make a (short) animation?

It might make sense to revisit some of the more expensive examples and reduce their cost too (until we find a way to run examples on tartarus, or some other way to make them less expensive).

**Comment by navidcy** [comment]
@francispoulin,@fra@fra

I agree! I'd like to plot movie of barotropic streamfunction or free surface and the relative vorticity. Perhaps total speed also. (I just plotted u and v because it was the easiest... just wanted to see whether things blow up or have grid-scale noise and what not.)

**Comment by navidcy** [comment]
> How long did this take you to run and what resolution did you use?

The animation above is a 10 year run from start. With a resolution of `Nx = 160`, `Ny = 240`, and `Nz = 50` it took bit less than an hour on the GPU (or something like that).

**Comment by hdrake** [comment]
Would it make sense to initialize with the analytical solution to the barotropic problem? Even if a simple Laplacian viscosity is a poor parameterization for eddy effects, it should dramatically accelerate the equilibration process, right?

**Comment by glwagner** [comment]
@navidcy we definitely want an example of this kind. It's a bit of work though, but perhaps we should bring this PR up to date?

**Comment by navidcy** [comment]
> @navidcy we definitely want an example of this kind. It's a bit of work though, but perhaps we should bring this PR up to date?

done

**Comment by francispoulin** [comment]
I updated the code so that it runs with the latest functions, and I also modified it to run with a lat-lon grid.  This is a result for the case with $N_z = 1$, which shows the voritcity on the left and the horizontal divergence on the right.  I thought it might be nice to have 2x2 plots where the lower row has the streamfunction and the speed.  

Also, I can run it with $N_z = 4$, since this example is from the MITgcm, and that case has four leves. 

What would people like to see?

https://user-images.githubusercontent.com/8239041/172967077-ff4d15bd-925d-4c44-af8e-5a3256e18d1f.mp4

**Comment by glwagner** [comment]
> I was looking for a reference to compare this to and found this paper by [Nadiga and Straub](https://www.sciencedirect.com/science/article/abs/pii/S1463500310000168).
> 
> They solve a one-layer Quasi-Geostrophic model so it will certainly have some differences, but if we have very low vertical resolution, I woudl expect that they should yield similar results. Sharing it in case it might be of interest.

Can you use the MITgcm documentation as a reference?

https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html

**Comment by francispoulin** [comment]
> > I was looking for a reference to compare this to and found this paper by [Nadiga and Straub](https://www.sciencedirect.com/science/article/abs/pii/S1463500310000168).
> > They solve a one-layer Quasi-Geostrophic model so it will certainly have some differences, but if we have very low vertical resolution, I woudl expect that they should yield similar results. Sharing it in case it might be of interest.
> 
> Can you use the MITgcm documentation as a reference?
> 
> https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html

I am using all of the same parameters at the [MITgcm Baroclinic Ocean Gyre example](https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html). 

I looked at it again and see that they have some results from year 100.  We can try and reproduce those, but it will take a while to get 100 years.

**Comment by glwagner** [comment]
> Also, I can run it with Nz = 4, since this example is from the MITgcm, and that case has four leves.

In my reading of the docs I see `Nz = 15`:

> As in tutorial [Barotropic Ocean Gyre](https://mitgcm.readthedocs.io/en/latest/examples/barotropic_gyre/barotropic_gyre.html#sec-eg-baro), a border row of land cells surrounds the ocean domain, so the full numerical grid size is 62 × 62 in the horizontal. The domain has 15 levels in the vertical, varying from Δz = 50 m deep in the surface layer to 190 m deep in the bottom layer, as shown by the faint red lines in [Figure 4.5](https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html#baroclinic-gyre-config).

From the first paragraph here: https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html

**Comment by francispoulin** [comment]
> > Also, I can run it with Nz = 4, since this example is from the MITgcm, and that case has four leves.
> 
> In my reading of the docs I see `Nz = 15`:
> 
> > As in tutorial [Barotropic Ocean Gyre](https://mitgcm.readthedocs.io/en/latest/examples/barotropic_gyre/barotropic_gyre.html#sec-eg-baro), a border row of land cells surrounds the ocean domain, so the full numerical grid size is 62 × 62 in the horizontal. The domain has 15 levels in the vertical, varying from Δz = 50 m deep in the surface layer to 190 m deep in the bottom layer, as shown by the faint red lines in [Figure 4.5](https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html#baroclinic-gyre-config).
> 
> From the first paragraph here: https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html. Is there another reference?

My mistake.  Yes, 15 points in the vertical. I will work on reproducing some of their diagnostics and confirm things compare.

**Comment by ali-ramadhan** [comment]
I updated the script and the simulation ran but I got a weird segfault during plotting. Did not try to replicate on another machine yet.

File: examples/double_gyre.jl [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+# ### Buoyancy relaxation
+
+@inline function buoyancy_relaxation(i, j, k, grid, clock, model_fields, p)
+   y = ynode(Center(), j, grid)
+   b = @inbounds model_fields.b[i, j, k]
+   return - 1 / p.λ * (b - p.Δb * y / p.Ly)
+end
+
+Fb = Forcing(buoyancy_relaxation, discrete_form = true, parameters = parameters)
```
Comment (ali-ramadhan) [inline examples/double_gyre.jl:87]: Seems like these would be cleaner if we used the continuous form?

Comment (navidcy) [reply examples/double_gyre.jl:87]: Agreed!

Comment (navidcy) [reply examples/double_gyre.jl]: Done in [00c2a69](https://github.com/CliMA/Oceananigans.jl/pull/1085/commits/00c2a698016e2f0dffe87b22e5ece64b23689648)

**Comment by navidcy** [comment]
[I moved this work from my fork to an Oceananigans branch.]

**Pull request closed without merging**